marlowe-chain-sync/src/Language/Marlowe/Runtime/ChainSync/Api.hs

```diff
@@ -0,0 +1,384 @@
+  | MinBound SlotNo           -- ^ The transaction is only valid after a specific slot.
+  | MaxBound SlotNo           -- ^ The transaction is only valid before a specific slot.
+  | MinMaxBound SlotNo SlotNo -- ^ The transaction is only valid between two slots.
+  deriving stock (Show, Read, Eq, Ord, Generic)
+  deriving anyclass (Binary)
+
+-- TODO invlude validator versions
+data Metadata
+  deriving stock (Show, Read, Eq, Ord, Generic)
+  deriving anyclass (Binary)
+
+-- | An input of a transaction.
+data TransactionInput = TransactionInput
+  { txId     :: !TxId             -- ^ The txId of the TransactionOutput this input consumes.
+  , txIx     :: !TxIx             -- ^ The txIx of the TransactionOutput this input consumes.
+  , redeemer :: !(Maybe Redeemer) -- ^ The script redeemer dataum for this input (if one was provided).
+  }
+  deriving stock (Show, Read, Eq, Ord, Generic)
+  deriving anyclass (Binary)
+
+-- | An output of a transaction.
+data TransactionOutput = TransactionOutput
+  { address   :: !Address           -- ^ The address that receives the assets of this output.
+  , assets    :: !Assets            -- ^ The assets this ouptut produces.
+  , datumHash :: !(Maybe DatumHash) -- ^ The hash of the script datum associated with this output.
+  , datum     :: !(Maybe Datum)     -- ^ The script datum associated with this output.
+  }
+  deriving stock (Show, Read, Eq, Ord, Generic)
+  deriving anyclass (Binary)
+
+-- | A script datum that is used to spend the output of a script tx.
+newtype Redeemer = Redeemer { unRedeemer :: Datum }
+  deriving stock (Show, Read, Eq, Ord, Generic)
+  deriving newtype (Binary)
+
+-- | A datum as a sum-of-products.
+data Datum
+  = Constr Integer [Datum]
+  | Map [(Datum, Datum)]
+  | List [Datum]
+  | I Integer
+  | B ByteString
+  deriving stock (Show, Read, Eq, Ord, Generic)
+  deriving anyclass (Binary)
+
+-- | Convert from Plutus.V1.Ledger.Api.Data to Datum
+fromPlutusData :: Plutus.Data -> Datum
+fromPlutusData (Plutus.Constr i dats) = Constr i $ fromPlutusData <$> dats
+fromPlutusData (Plutus.Map m)         = Map $ bimap fromPlutusData fromPlutusData <$> m
+fromPlutusData (Plutus.List dats)     = List $ fromPlutusData <$> dats
+fromPlutusData (Plutus.I i)           = I i
+fromPlutusData (Plutus.B b)           = B b
+
+-- | A collection of assets transferred by a trasaction output.
+data Assets = Assets
+  { ada    :: !Lovelace -- ^ The ADA sent by the tx output.
+  , tokens :: !Tokens   -- ^ Additional tokens sent by the tx output.
+  }
+  deriving stock (Show, Read, Eq, Ord, Generic)
+  deriving anyclass (Binary)
+
+-- | A collection of token quantities by their asset ID.
+newtype Tokens = Tokens { unTokens :: Map AssetId Quantity }
+  deriving stock (Show, Read, Eq, Ord, Generic)
+  deriving newtype (Binary, Semigroup, Monoid)
+
+-- | A newtype wrapper for parsing base 16 strings as byte strings.
+newtype Base16 = Base16 { unBase16 :: ByteString }
+
+instance Show Base16 where
```

Comment:
It's unfortunate that it would be awkward to use `Cardano.Api.SerialiseToRawBytesHex`..